packages/catalog-process/src/services/catalogService.ts

```diff
@@ -463,6 +464,33 @@ export function catalogServiceBuilder(
         })
       );
     },
+
+    async getEServiceById(
+      eServiceId: EServiceId,
+      authData: AuthData
+    ): Promise<EService> {
+      const eService = await retrieveEService(eServiceId, readModelService);
+
+      if (
+        authData.organizationId === eService.data.producerId &&
+        (authData.userRoles.includes(userRoles.ADMIN_ROLE) ||
+          authData.userRoles.includes(userRoles.API_ROLE))
+      ) {
+        return eService.data;
+      }
+      const eServiceWithoutDraft: EService = {
+        ...eService.data,
+        descriptors: eService.data.descriptors.filter(
+          (d) => d.state !== descriptorState.draft
+        ),
+      };
+
+      if (eServiceWithoutDraft.descriptors.length === 0) {
+        throw eServiceNotFound(eServiceId);
+      }
```

Comment:
This code will be needed in other routes, I suggest to extract it in a dedicated function.
Also you could check the descriptors before creating `eServiceWithoutDraft`, because it is not needed in case there's only a `DRAFT` descriptor

Reply:
For the first comment: I implemented the other PRs you are taking about, and I have seen this is not reusable. Each endpoint has some customization to this check.

For the second comment, I am not sure about what you are suggesting. The logic I wrote is to summarize the following restrictions, without being too verbose:
- eService with no descriptor: error `404`
- eService with only one descriptor, which is `draft`: error `404`
- eService with `draft` and `non-draft` descriptor: filter out the draft ones.

Reply:
the same checks are performed on `getEServiceDocumentById`, I agree on creating a separate function (Scala code ref [here](https://github.com/pagopa/interop-be-catalog-process/blob/1.0.x/src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala#L401))
Plus, it helps delineating the specific business logic, improving readability and maintenability

Reply:
We already discussed with @AsterITA what could be a good tradeoff between readability and reusability. We reached the solution of reusing the function `isUserAllowedToSeeDraft` (the code shown on top of this conversation is outdated) and keep the business logic as optimized as we can for the specific use cases. References: [this](https://github.com/pagopa/interop-be-monorepo/pull/192/files#diff-77a5bad1f3b877e7bfc968a9b423c16649fca358d54286d743fe34fa555564ceR234-R248) and [this](https://github.com/pagopa/interop-be-monorepo/pull/197/files#diff-77a5bad1f3b877e7bfc968a9b423c16649fca358d54286d743fe34fa555564ceR361-R367).

The problem of the solution you suggest is that the reuse of the entire `applyVisibilityToEService` in the Scala code, brings other tradeoffs in the business logic: [this](https://github.com/pagopa/interop-be-catalog-process/blob/1.0.x/src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala#L403-L404) part is not optimal for me, because the eService gets "cleaned", and after that if we still find the document in it, we return it. Instead, in this PR the check is specifically on the descriptor state, so we get a better idea of why we are not returning the document.

Reply:
Update: after another discussion, we followed the approach of using the `applyVisibilityToEService` function.